### ISSUE TITLE
Cookies for OIDC filter are being set for wrong domain

### ISSUE THREAD

**Issue by abinet** [zalando/skipper#1909]
We are experiencing an issue with skipper running in multiple environments dev and test and oidc cookie.

Skipper is creating two unique cookies per host like: 

host: dev.applications.example.com
skipperOauthOidc2c6815ff-a
skipperOauthOidc2c6815ff-b

host: test.applications.example.com
skipperOauthOidc84ee2c5e-a
skipperOauthOidc84ee2c5e-b

The problem is that skipper stores both cookie sets under parent host:
host: applications.example.com
skipperOauthOidc2c6815ff-a
skipperOauthOidc2c6815ff-b
skipperOauthOidc84ee2c5e-a
skipperOauthOidc84ee2c5e-b

And this leads to an 400 Bad Request "Request Header Or Cookie Too Large"

I suppose here is the root cause:
https://github.com/zalando/skipper/blame/master/filters/auth/oidc.go#L374
Not sure why it is needed to remove one domain level here.

**Comment by szuecs** [comment]
@abinet we suppose that you have likely multiple applications that you want to have SSO for.
For example you have `a.example.org` and `b.example.org` and you want to use the same identity and cookie for both.
So if you would use:
`applications.dev.example.com`
`applications.test.example.com`
`applications.example.com`

it should also work.

**Comment by abinet** [comment]
Thank you @szuecs  for the details and workaround. The subdomain removal make sense when you put a single skipper instance in front of multiple environments. However it does not work for cases like ours and changing the dns schema is quite complicated in our case as well. What would you say if we make this behavior configurable introducing either instance setting or filter parameter like  "extract-domain-level" with default value 2?

**Comment by AlexanderYastrebov** [comment]
> What would you say if we make this behavior configurable introducing either instance setting or filter parameter like "extract-domain-level" with default value 2?

`oauthOidc*` filters already have quite bloated interface it would be challenging to cram a new parameter :(

**Comment by AlexanderYastrebov** [comment]
We've have discussed the issue with @szuecs and came to the following points.

* it makes sense to have cookie domain level configurable in general as per https://github.com/zalando/skipper/issues/1909#issuecomment-984620970
* the configuration should be per filter, not global. Having it global makes it obscure/non-consistent and postpones the problem until when different domain levels are needed per route
* the configuration should be a new (third) optional parameter of `oauthOidc*` filters. The documentation for optional parameters should be updated to mention default empty value. We thought about adding new `oauthOidcConfig(...)` filter but that feels like an overkill. Users that require to configure domain level would need to specify empty values for preceding optional parameters if they are not used:
    ```
    oauthOidcAllClaims("https://oidc-provider.example.com", "client_id", "client_secret",
        "http://target.example.com/subpath/callback", "email profile", "name email picture",
        "", "", 1)
    ```
* the configured value should be non-negative integer (eskip does not support negative values #194 and there was no need so far) specifying how many subdomains to remove from the request domain, e.g. 1 is the default for backward compatibility, 0 is to use request domain as is, etc.

**Comment by abinet** [comment]
@szuecs @AlexanderYastrebov Just small clarification. Current logic is to remove subdomains only starting from the third level:

localhost -> localhost
example.com -> example.com
www.example.com -> example.com

If you want to keep this logic for backward compatibility reason it would mean that all the removal logic applies for hosts with starting with 3th domain level

| Host | DomainLevelToRemove | Result |
|--|--|--|
| localhost | 1 | localhost |
| example.com | 1 | example.com | 
| example.com | 2 | example.com |
| www.example.com | 0 | www.example.com |
| www.example.com | 1 | example.com |
| www.example.com | 2 | com |

**Comment by AlexanderYastrebov** [comment]
> all the removal logic applies for hosts with starting with 3th domain level

yes. But I think it is wrong to set cookie for the TLD https://stackoverflow.com/questions/53848070/can-cookie-be-set-against-top-level-domain-like-com so this

| Host | DomainLevelToRemove | Result |
|--|--|--|
www.example.com | 2 | com

is not correct. 

I would prefer that code simply removes as many subdomains as configured and rely on the user to configure proper value for the their domain but we have to stay compatible so we would need to keep and document level check logic.

We would need to check something like:
```go
levels := 1 + strings.Count(h, ".")
if levels - removeSubdomains < 2 {
    return h 
}
```
And the `extractDomainFromHost("www.example.com", 2)` should probably return `"www.example.com"` intact.


> In addition I've just discovered that GrantLogout, GrantCallback filters are using same method to extract host name.

We can keep the status quo for grant filters like
```go
Domain:   extractDomainFromHost(host, 1),
```
and decide if level needs to be configurable in the same manner later. 

>  So, it is probably make sense to introduce global setting which can be used by all filters.

For OIDC global makes it obscure/non-consistent as all other configs are per-filter and postpones the problem until when different domain levels are needed per route. For grant I am not sure yet, maybe it will be global option like other grant options but I think we can postpone grant changes until necessary.

**Comment by AlexanderYastrebov** [comment]
@abinet Thank you for your contribution.

Fixed by #1923
